### PR TITLE
added concurrentDictionary to store httpContext for requests.

### PR DESCRIPTION
**Pull request by amirmohammadzakeri**
با این تغییر،
مموری کمتری مصرف می‌شود.
در واقع بیشترین مصرف مموری به دلیل ایجاد کردن grain های بسیار بود.

### CODE REVIEW

**Comment by amirmohammadzakeri** [comment]
SingleGrainPerSilo:
  MaxMemoryUsage: 45MB
          /\      |‾‾| /‾‾/   /‾‾/
     /\  /  \     |  |/  /   /  /
    /  \/    \    |     (   /   ‾‾\
   /          \   |  |\  \ |  (‾)  |
  / __________ \  |__| \__\ \_____/ .io

  execution: local
     script: script.js
     output: -

  scenarios: (100.00%) 1 scenario, 2 max VUs, 10m30s max duration (incl. graceful stop):
           * default: 100000 iterations shared among 2 VUs (maxDuration: 10m0s, gracefulStop: 30s)


     data_received..................: 13 MB  626 kB/s
     data_sent......................: 8.5 MB 426 kB/s
     http_req_blocked...............: avg=1.73µs   min=0s med=0s       max=16.99ms p(90)=0s       p(95)=0s
     http_req_connecting............: avg=10ns     min=0s med=0s       max=1ms     p(90)=0s       p(95)=0s
     http_req_duration..............: avg=342.58µs min=0s med=504.49µs max=32ms    p(90)=682.7µs  p(95)=999.8µs
       { expected_response:true }...: avg=342.58µs min=0s med=504.49µs max=32ms    p(90)=682.7µs  p(95)=999.8µs
     http_req_failed................: 0.00%  ✓ 0           ✗ 100000
     http_req_receiving.............: avg=30.64µs  min=0s med=0s       max=8.79ms  p(90)=0s       p(95)=475.81µs
     http_req_sending...............: avg=9.4µs    min=0s med=0s       max=5.58ms  p(90)=0s       p(95)=0s
     http_req_tls_handshaking.......: avg=0s       min=0s med=0s       max=0s      p(90)=0s       p(95)=0s
     http_req_waiting...............: avg=302.54µs min=0s med=0s       max=31.61ms p(90)=649.61µs p(95)=999.2µs
     http_reqs......................: 100000 5007.879222/s
     iteration_duration.............: avg=390.82µs min=0s med=516.29µs max=32ms    p(90)=764µs    p(95)=1ms
     iterations.....................: 100000 5007.879222/s
     vus............................: 2      min=2         max=2
     vus_max........................: 2      min=2         max=2

SingleGrainPerRequest:
  MaxMemoryUsage: 500MB
          /\      |‾‾| /‾‾/   /‾‾/
     /\  /  \     |  |/  /   /  /
    /  \/    \    |     (   /   ‾‾\
   /          \   |  |\  \ |  (‾)  |
  / __________ \  |__| \__\ \_____/ .io

  execution: local
     script: script.js
     output: -

  scenarios: (100.00%) 1 scenario, 2 max VUs, 10m30s max duration (incl. graceful stop):
           * default: 100000 iterations shared among 2 VUs (maxDuration: 10m0s, gracefulStop: 30s)


     data_received..................: 13 MB  462 kB/s
     data_sent......................: 8.5 MB 314 kB/s
     http_req_blocked...............: avg=2.57µs   min=0s med=0s       max=24.93ms  p(90)=0s      p(95)=0s
     http_req_connecting............: avg=36ns     min=0s med=0s       max=2.24ms   p(90)=0s      p(95)=0s
     http_req_duration..............: avg=457.05µs min=0s med=506.6µs  max=78.96ms  p(90)=989µs   p(95)=1.01ms
       { expected_response:true }...: avg=457.05µs min=0s med=506.6µs  max=78.96ms  p(90)=989µs   p(95)=1.01ms
     http_req_failed................: 0.00%  ✓ 0           ✗ 100000
     http_req_receiving.............: avg=31.89µs  min=0s med=0s       max=6.21ms   p(90)=0s      p(95)=362.83µs
     http_req_sending...............: avg=11.22µs  min=0s med=0s       max=11.99ms  p(90)=0s      p(95)=0s
     http_req_tls_handshaking.......: avg=0s       min=0s med=0s       max=0s       p(90)=0s      p(95)=0s
     http_req_waiting...............: avg=413.93µs min=0s med=505.29µs max=78.96ms  p(90)=831.3µs p(95)=1ms
     http_reqs......................: 100000 3698.154267/s
     iteration_duration.............: avg=524.79µs min=0s med=519.7µs  max=104.89ms p(90)=999.6µs p(95)=1.21ms
     iterations.....................: 100000 3698.154267/s
     vus............................: 2      min=2         max=2
     vus_max........................: 2      min=2         max=2

File: src/Services/StreamProcessing/StreamProcessing/HttpListener/HttpListenerResponseLocalGrain.cs [hunk 1]
```diff
@@ -39,23 +43,26 @@ public HttpListenerResponseLocalGrain(IPluginOutputCaller pluginOutputCaller)
...
-            response.Close();            
+            foreach (var header in responseTuple.Headers)
+            {
+                response.AddHeader(header.Key, header.Value);
+            }
+
+            // returning response 
+            if (responseTuple.ContentBytes is not null)
+            {
+                response.Close(responseTuple.ContentBytes,false);
```
Comment (AhmadAfsharinejad) [inline src/Services/StreamProcessing/StreamProcessing/HttpListener/HttpListenerResponseLocalGrain.cs]: لطفا همیشه قبل از سیو ،  ریفرمت رو بزن تا اینا اوکی شند

File: src/Services/StreamProcessing/StreamProcessing/HttpListener/HttpListenerResponseLocalGrain.cs [hunk 2]
```diff
@@ -16,8 +17,8 @@ namespace StreamProcessing.HttpListener;
 internal sealed class HttpListenerResponseLocalGrain : Grain, IHttpListenerResponseLocalGrain
 {
     private readonly IPluginOutputCaller _pluginOutputCaller;
-    private HttpListenerContext? _httpListenerContext;
-
+    private readonly ConcurrentDictionary<Guid,HttpListenerContext> _httpListenerContextDictionary = new();
```
Comment (AhmadAfsharinejad) [inline src/Services/StreamProcessing/StreamProcessing/HttpListener/HttpListenerResponseLocalGrain.cs]: اگر از لیست استفاده شه مشکلی پیش میاد؟|
با توجه به اینکه گرین ها thread safe هستند و ما از اتریبیوت readonly استفاده نکردیم

File: src/Services/StreamProcessing/StreamProcessing/HttpListener/HttpListenerResponseLocalGrain.cs (hunk 1 above)
Comment (amirmohammadzakeri) [reply src/Services/StreamProcessing/StreamProcessing/HttpListener/HttpListenerResponseLocalGrain.cs]: Done.

File: src/Services/StreamProcessing/StreamProcessing/HttpListener/HttpListenerResponseLocalGrain.cs (hunk 2 above)
Comment (amirmohammadzakeri) [reply src/Services/StreamProcessing/StreamProcessing/HttpListener/HttpListenerResponseLocalGrain.cs]: به جای این کار گرین رو به Reentrant تبدیل کردیم.

**Pull request merged**